src/app/clusters/group-key-mgmt-server/group-key-mgmt-server.cpp

```diff
@@ -0,0 +1,410 @@
+/**
+ *
+ *    Copyright (c) 2020 Project CHIP Authors
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#include <app/CommandHandler.h>
+#include <app/common/gen/att-storage.h>
+#include <app/common/gen/attribute-id.h>
+#include <app/common/gen/attribute-type.h>
+#include <app/common/gen/attributes/Accessors.h>
+#include <app/common/gen/cluster-id.h>
+#include <app/common/gen/command-id.h>
+#include <app/util/af.h>
+#include <map>
+#include <set>
+#include <string.h>
+#include <tuple>
+#include <utility>
+
+constexpr size_t GROUP_KEY_ROOT_SIZE  = 16;
+constexpr size_t GROUP_KEY_EPOCH_SIZE = 8;
+constexpr uint8_t GROUP_KEY_SET_SIZE  = 3;
```

Comment:
It is clear that the spec intends there to be 3 epoch keys active at any given time.  It is not clear, however, that one of the four possible epoch index values is impossible or will never be seen.

I believe there are two ways to deal with this:
* infer epoch index as the PR currently does (albeit incorrectly)
* accept all four possible epoch indices and actually hold all four in memory

I doubt we want to waste memory so maybe it is good to have optimized and explicitly only ever hold three keys per KeySetIndex.  But something needs to be done here to correctly deal with this.

Again, GroupKeyIndex != GroupKeySetIndex, and it is not possible to infer GroupKeyEpochIndex based upon the time of key-set operation.